### PR TITLE
Event Checkin and Event Edit Fragments (EXPOSUREAPP-5743/5953)

### PR DESCRIPTION
**Pull request by kolyaopahle**
### Description
This PR contains the checkin and edit fragments for the event registration feature

### CODE REVIEW

**Comment by mtwalli** [comment]
1- Duration text is very close to the text 
device-2021-03-26-161702

2- Background is scrolling but not the content 
https://user-images.githubusercontent.com/25054729/112653699-fa211f00-8e4e-11eb-976a-6286595e2655.mp4

**Comment by mtwalli** [comment]
2- Background is scrolling but not the content 
https://user-images.githubusercontent.com/25054729/112653699-fa211f00-8e4e-11eb-976a-6286595e2655.mp4

**Comment by mtwalli** [comment]
Light mode looks odd:
device-2021-03-26-162118

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/EventRegistrationUIModule.kt [hunk 1]
```diff
@@ -22,6 +24,9 @@ internal abstract class EventRegistrationUIModule {
     @ContributesAndroidInjector(modules = [ConfirmCheckInModule::class])
     abstract fun confirmCheckInFragment(): ConfirmCheckInFragment
 
+    @ContributesAndroidInjector(modules = [EditCheckInModule::class])
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/EventRegistrationUIModule.kt:27]: EditCheckIn`Fragment`Module?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/attendee/confirm/ConfirmCheckInViewModel.kt [hunk 2]
```diff
@@ -62,8 +110,30 @@ class ConfirmCheckInViewModel @AssistedInject constructor(
...
+        private val traceLocation: TraceLocation,
+        private val checkInEndOffset: Duration,
+        val createJournalEntry: Boolean,
+        val eventInPastVisible: Boolean,
+        val eventInFutureVisible: Boolean
+    ) {
+        val description get() = traceLocation.description
+        val typeRes get() = mapTraceLocationToTitleRes(traceLocation.type)
+        val address get() = traceLocation.address
+        val checkInEnd get() = checkInEndOffset.toReadableDuration(suffix = "Std")
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/attendee/confirm/ConfirmCheckInViewModel.kt:127]: should not `Std` be localised ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/attendee/edit/EditCheckInFragment.kt [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+                viewModel.onSaveClicked()
+            }
+        }
+
+        viewModel.events.observe2(this) { navEvent ->
+            when (navEvent) {
+                EditCheckInNavigation.BackNavigation -> popBackStack()
+                EditCheckInNavigation.ConfirmNavigation -> {
+                    popBackStack()
+                }
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/attendee/edit/EditCheckInFragment.kt:71]: `{}` not needed

File: Corona-Warn-App/src/main/res/layout/fragment_confirm_check_in.xml [hunk 4]
```diff
@@ -1,54 +1,231 @@
...
+            android:id="@+id/coordinator_layout"
+            android:layout_width="0dp"
+            android:layout_height="0dp"
+            android:layout_marginBottom="@dimen/spacing_small"
+            android:nestedScrollingEnabled="true"
+            app:layout_constraintBottom_toTopOf="@id/confirm_checkin_confirm_button"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toTopOf="parent"
+            tools:context=".CollapsingToolbar">
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/res/layout/fragment_confirm_check_in.xml:21]: tools:context=".CollapsingToolbar" ? should not be the fragment?

File: Corona-Warn-App/src/main/res/layout/fragment_confirm_check_in.xml [hunk 5]
```diff
@@ -1,54 +1,231 @@
...
+                    <LinearLayout
+                        android:id="@+id/confirm_checkin_settings_card"
+                        style="@style/Card"
+                        android:layout_width="match_parent"
+                        android:layout_height="wrap_content"
+                        android:layout_marginHorizontal="@dimen/spacing_normal"
+                        android:layout_marginTop="@dimen/spacing_tiny"
+                        android:orientation="vertical">
+
+                        <androidx.appcompat.widget.SwitchCompat
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/res/layout/fragment_confirm_check_in.xml:172]: Please use `MaterialSwitch`

File: Corona-Warn-App/src/main/res/layout/fragment_edit_check_in.xml [hunk 6]
```diff
@@ -0,0 +1,246 @@
+<?xml version="1.0" encoding="utf-8"?>
+
+<layout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools">
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/res/layout/fragment_edit_check_in.xml:5]: Please remove `<layout>` tag?

File: Corona-Warn-App/src/main/res/layout/fragment_edit_check_in.xml [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+                            android:text="@string/edit_checkin_duration_edit_hint_card_text" />
+                    </LinearLayout>
+
+                </LinearLayout>
+
+            </androidx.core.widget.NestedScrollView>
+
+        </androidx.coordinatorlayout.widget.CoordinatorLayout>
+
+        <android.widget.Button
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/res/layout/fragment_edit_check_in.xml:231]: why` <android.widget.` ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/attendee/edit/EditCheckInFragment.kt [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+                editCheckinInfoCardTitle.text = uiState.description
+                editCheckinInfoCardAddress.text = uiState.address
+
+                editCheckinEditCardCheckinDate.text = uiState.checkInStartDate
+                editCheckinEditCardCheckinTime.text = uiState.checkInStartTime
+
+                editCheckinEditCardCheckoutDate.text = uiState.checkInEndDate
+                editCheckinEditCardCheckoutTime.text = uiState.checkInEndTime
+
+                editCheckinDurationEditHintCard.setGone(!uiState.diaryWarningVisible)
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/attendee/edit/EditCheckInFragment.kt:87]: `isGone` from `ktx` ? hopefully we can get rid of DataBinding and its binding adapters

File: Corona-Warn-App/src/main/res/layout/fragment_confirm_check_in.xml [hunk 9]
```diff
@@ -1,54 +1,231 @@
 <?xml version="1.0" encoding="utf-8"?>
-<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+
+<layout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/res/layout/fragment_confirm_check_in.xml:3]: Please remove `<layout>` tag ?

**Comment by ChrisLoes** [comment]
Hey @kolyaopahle,

thanks for implementing this! Is there any public documentation/information about the concept of this feature? I've only found this https://github.com/corona-warn-app/cwa-documentation/blob/master/event_registration.md#interoperability-with-other-contact-tracing-apps-draft - but it's only a draft...

thank you!

**Comment by kolyaopahle** [comment]
Hi @ChrisLoes i think what you found is the most up to date publicly available information, not all parts of the specification are currently finally defined as the feature is still in active development, if you have any specific questions feel free to reach out and i will try to answer as good as i can 🙂

**Comment by ChrisLoes** [comment]
@kolyaopahle thanks for your quick response. 

It looks like the cwa will be able to scan qr codes also from other presence tracing apps if they implement the suggested url pattern specified in the draft. 

Do you know how the encoded_signed_trace_location will be generated and how other apps like luca or corona-anmeldung.de can get or generate it?

**Comment by kolyaopahle** [comment]
@ChrisLoes unfortunately this is a part of the proposal that is still receiving some changes, as of right now the idea is that the cwa backend returns these signed location upon request on the appropriate api endpoint, but there are a few changes planned as of last Friday that should develop in the coming week and (if implemented) will substantially change how event creation and checkin is handled. I'll mention you here if i have any news on the topic🙂

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/attendee/edit/EditCheckInViewModel.kt [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+    }
+
+    fun onEndTimeChanged(event: DateTimePickerEvent) {
+        val endDateTime = checkInEndTime.value?.toDateTime()
+
+        when (event) {
+            is DateTimePickerEvent.TimePickerEvent ->
+                checkInEndTime.value = endDateTime?.withTime(event.localTime)?.toInstant()
+            is DateTimePickerEvent.DatePickerEvent ->
+                checkInEndTime.value = endDateTime?.withDate(event.localDate)?.toInstant()
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/attendee/edit/EditCheckInViewModel.kt:104]: No validation required here such as `endDate > startDate` ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/organizer/category/adapter/category/TraceLocationCategory.kt [hunk 11]
```diff
@@ -106,3 +106,11 @@ val traceLocationCategories = listOf(
         R.string.tracelocation_organizer_category_other_event_title
     )
 )
+
+fun mapTraceLocationToTitleRes(type: TraceLocationOuterClass.TraceLocationType) =
+    mapTraceLocationToTitleRes(type.number)
+
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/organizer/category/adapter/category/TraceLocationCategory.kt:112]: `@StringRes` ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/attendee/edit/EditCheckInViewModel.kt [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+        ): EditCheckInViewModel
+    }
+
+    data class UiState(
+        private val checkIn: CheckIn,
+        private val checkInStartInstant: Instant,
+        private val checkInEndInstant: Instant
+    ) {
+        val description: String get() = checkIn.description
+        val typeRes: Int get() = mapTraceLocationToTitleRes(checkIn.type)
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/attendee/edit/EditCheckInViewModel.kt:139]: `@StringRes` ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/organizer/category/adapter/category/TraceLocationCategory.kt [hunk 13]
```diff
@@ -106,3 +106,11 @@ val traceLocationCategories = listOf(
         R.string.tracelocation_organizer_category_other_event_title
     )
 )
+
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/organizer/category/adapter/category/TraceLocationCategory.kt:109]: `@StringRes` ?

File: Corona-Warn-App/src/main/res/layout/fragment_confirm_check_in.xml [hunk 14]
```diff
@@ -1,54 +1,231 @@
...
+                    android:nestedScrollingEnabled="true"
+                    app:collapsedTitleTextAppearance="@style/headline5"
+                    app:expandedTitleTextAppearance="@style/headline5"
+                    app:layout_scrollFlags="scroll|exitUntilCollapsed">
+
+                    <ImageView
+                        android:id="@+id/expandedImage"
+                        android:layout_width="match_parent"
+                        android:layout_height="300dp"
+                        android:src="@drawable/trace_location_view_cardhighlight_gradient_all_corners"
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/res/layout/fragment_confirm_check_in.xml:41]: `srcCompat` ?

File: Corona-Warn-App/src/main/res/layout/fragment_confirm_check_in.xml [hunk 15]
```diff
@@ -1,54 +1,231 @@
...
+                            android:layout_width="match_parent"
+                            android:layout_height="match_parent"
+                            android:gravity="center_vertical"
+                            android:orientation="horizontal">
+
+                            <ImageView
+                                android:layout_width="match_parent"
+                                android:layout_height="wrap_content"
+                                android:layout_marginEnd="72dp"
+                                android:src="@drawable/ic_cwa_logo" />
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/res/layout/fragment_confirm_check_in.xml:67]: `srcCompat` ?

File: Corona-Warn-App/src/main/res/layout/fragment_confirm_check_in.xml [hunk 16]
```diff
@@ -1,54 +1,231 @@
...
+
+                    </LinearLayout>
+
+                </LinearLayout>
+
+            </androidx.core.widget.NestedScrollView>
+
+        </androidx.coordinatorlayout.widget.CoordinatorLayout>
+
+        <android.widget.Button
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/res/layout/fragment_confirm_check_in.xml:216]: Please don't use` <android.widget.Button>`? FYI https://github.com/corona-warn-app/cwa-app-android/pull/2700

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/EventRegistrationUIModule.kt (hunk 1 above)
Comment (kolyaopahle) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/EventRegistrationUIModule.kt:27]: id like to keep the pattern we used everywhere else

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/attendee/edit/EditCheckInViewModel.kt (hunk 12 above)
Comment (kolyaopahle) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/attendee/edit/EditCheckInViewModel.kt:139]: i cannot annotate a getter afaik

Comment (mtwalli) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/attendee/edit/EditCheckInViewModel.kt:139]: `@get:StringRes ` not a blocker for me

Comment (kolyaopahle) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/attendee/edit/EditCheckInViewModel.kt]: ah right, completely forgot about that :D

**Comment by axelherbstreith** [comment]
@mtwalli UI is now updated 👍 
The current implementation allows scrolling when texts are to large for the screen size.

https://user-images.githubusercontent.com/75120552/112857413-1d4b0900-90b1-11eb-9a79-ca30d909454a.mp4

**Comment by mtwalli** [comment]
@axelherbstreith  Thanks for fixing the scrolling issue, @kolyaopahle  I ca not see `Std` label anymore please double check that?

### PR
Screenshot 2021-03-29 at 18 09 37

### Design
Screenshot 2021-03-29 at 18 12 47

**Comment by mtwalli** [comment]
Please check card elevations in both screens, as per design the elevation should not be that high
| device-2021-03-29-181859 | 
Screenshot 2021-03-29 at 18 19 33|

**Comment by kolyaopahle** [comment]
its the elevation we always used for cards in the app, im not sure we should change this just because of the design

**Comment by mtwalli** [comment]
Looks nice! , just one small comment.
The duration text and the title next to it should be top aligned as per design
Screenshot 2021-03-30 at 10 49 28

**Review by mtwalli — DISMISSED** [review]
LGTM

**Review by mtwalli — DISMISSED** [review]

**Review by chris-cwa — APPROVED** [review]

**Review by mtwalli — APPROVED** [review]

**Comment by ChrisLoes** [comment]
> @ChrisLoes unfortunately this is a part of the proposal that is still receiving some changes, as of right now the idea is that the cwa backend returns these signed location upon request on the appropriate api endpoint, but there are a few changes planned as of last Friday that should develop in the coming week and (if implemented) will substantially change how event creation and checkin is handled. I'll mention you here if i have any news on the topic🙂



> @ChrisLoes unfortunately this is a part of the proposal that is still receiving some changes, as of right now the idea is that the cwa backend returns these signed location upon request on the appropriate api endpoint, but there are a few changes planned as of last Friday that should develop in the coming week and (if implemented) will substantially change how event creation and checkin is handled. I'll mention you here if i have any news on the topic🙂

Hi @kolyaopahle ,

are there any updates/news regarding the event creation and how other presence apps can integrate it?

**Comment by mlenkeit** [comment]
@ChrisLoes https://github.com/corona-warn-app/cwa-documentation/blob/master/event_registration.md is the only available description at this point. It _is_ still labeled as draft but it is effectively representing what's currently implemented.

**Comment by ChrisLoes** [comment]
Thank you @mlenkeit ! Glad to see that there is no server connection needed anymore! Are there any release date plans or beta versions of cwa available to test the integration from other presence apps?

**Comment by ChrisLoes** [comment]
@mlenkeit Do you know how other tracing apps should set the properties for a valid CrowdNotifierData message?

message CrowdNotifierData {
  uint32 version = 1;
  bytes publicKey = 2; **-> who's public key / how to get or generate?**
  bytes cryptographicSeed = 3; **-> which seed?**
  uint32 type = 4; // exact semantic tbd
}

It would be very nice if there were some more information available how other presence-apps could generate the right data for cwa? :)

**Pull request merged**